samples/bluetooth/peripheral_hids_keyboard/src/main.c

```diff
@@ -0,0 +1,629 @@
+	hids_init_obj.info.flags = (HIDS_REMOTE_WAKE |
+			HIDS_NORMALLY_CONNECTABLE);
+
+	hids_inp_rep = &hids_init_obj.inp_rep_group_init.reports[0];
+	hids_inp_rep->size = INPUT_REPORT_KEYS_MAX_LEN;
+	hids_inp_rep->id = INPUT_REP_KEYS_REF_ID;
+	hids_init_obj.inp_rep_group_init.cnt++;
+
+	hids_outp_rep = &hids_init_obj.outp_rep_group_init.reports[0];
+	hids_outp_rep->size = OUTPUT_REPORT_MAX_LEN;
+	hids_outp_rep->id = OUTPUT_REP_REF_ID;
+	hids_outp_rep->handler = hids_outp_rep_handler;
+	hids_init_obj.outp_rep_group_init.cnt++;
+
+	hids_init_obj.is_kb = true;
+	hids_init_obj.boot_kb_outp_rep_handler = hids_boot_kb_outp_rep_handler;
+	hids_init_obj.pm_evt_handler = hids_pm_evt_handler;
+
+	err = hids_init(&hids_obj, &hids_init_obj);
+	__ASSERT(err == 0, "HIDS initialization failed\n");
+}
+
+
+static void bt_ready(int err)
+{
+	if (err) {
+		printk("Bluetooth init failed (err %d)\n", err);
+		return;
+	}
+
+	printk("Bluetooth initialized\n");
+
+	bas_init();
+	hid_init();
+
+	if (IS_ENABLED(CONFIG_SETTINGS)) {
+		settings_load();
+	}
+
+	advertising_start();
+}
+
+
+#if !defined(CONFIG_BT_GATT_HIDS_SECURITY_LEVEL_LOW) && \
+	defined(CONFIG_BT_GATT_HIDS_SECURITY_ENABLED)
+static void auth_passkey_display(struct bt_conn *conn, unsigned int passkey)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	printk("Passkey for %s: %06u\n", addr, passkey);
+}
+
+
+static void auth_cancel(struct bt_conn *conn)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	printk("Pairing cancelled: %s\n", addr);
+}
+
+
+static void auth_done(struct bt_conn *conn)
+{
+	printk("%s()\n", __func__);
+	bt_conn_auth_pairing_confirm(conn);
+}
+
+
+static struct bt_conn_auth_cb conn_auth_callbacks = {
+	.passkey_display = auth_passkey_display,
+	.passkey_entry = NULL,
+	.cancel = auth_cancel,
+	.pairing_confirm = auth_done,
+};
+#else
+static struct bt_conn_auth_cb conn_auth_callbacks;
+#endif
+
+/**@brief Function for key press simulation on selected connection
+ *
+ * This function sends key press and release.
+ *
+ * In case a pattern of 'cofFEe' is the p_key_pattern, with pattern_len as 6 and
+ * pattern_offset as 0, the notifications as observed on the peer side would be
+ * 1>    'c', 'o', 'f', 'F', 'E', 'e'
+ * 2>    -  , 'o', 'f', 'F', 'E', 'e'
+ * 3>    -  ,   -, 'f', 'F', 'E', 'e'
+ * 4>    -  ,   -,   -, 'F', 'E', 'e'
+ * 5>    -  ,   -,   -,   -, 'E', 'e'
+ * 6>    -  ,   -,   -,   -,   -, 'e'
+ * 7>    -  ,   -,   -,   -,   -,  -
+ * Here, '-' refers to release, 'c' refers to the key character being
+ * transmitted.
+ * Therefore 7 notifications will be sent.
+ *
+ * @param string    Key codes to be sent.
+ * @param len       Number of buttons to be sent.
+ * @param conn      Connection handler.
+ * @param boot_mode Information if boot mode protocol is selected.
+ *
+ * @return 0 on success or negative error code.
+ */
+int key_press_conn_send(u8_t *string,
```

Comment:
Isn't this queue function a bit too advanced for a HID keyboard? It should suffice to send a small string directly on button press instead.

Reply:
No - it is not, and it is not so simple. And I have even added quite big comment above why it is so complicated. Please review it. Basically send a string to PC from a keyboard is composed from pressing and RELEASING a buttons. You may not leave it pressed as you would get auto repetition.

Reply:
My point is that this queue is too advanced, and holds too many restriction on the data format.

Reply:
What kind of restriction it holds?

Reply:
See my second reply. These queue functions make the application harder to use as a real keyboard or HID pen application, as you would need to remove the queue related logic and add functionality that sends/releases per button press (ie: interrupt on rising and falling). Why not remove logic that is very likely to be removed by developers anyway?

Reply:
Ok.. I get your point. This is simple example. Managment of the buttons that are pressed, the buttons in the report that are released and remembering what of the buttons release was already sent. Packing it all dynamically is quite complex task. We have it solved in Smart Remote sample. And there is 1k line hid_state module to manage that.

This example was implemented as a side work it only to test some of the HID central features and it did the work required for me. I spend half a day today trying to solve the challenge you mentioned here, but the code is more and more complex.

Currently I would vote to leave it that way as a sample base. Please feel free to improve it in the future or use Smart Remote as a base for more advanced product.